### PR TITLE
Add PrecompiledToken

### PR DESCRIPTION
**Pull request by ThunderDeliverer**
Added PrecompiledToken example. The example introduces the precompiled
and predeployed smart contracts of the Acala EVM+ as well as the ADDRESS
utility of @acala-network/contracts dependency.

### CODE REVIEW

File: precompiled-token/src/getACAinfo.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import { use } from 'chai';
+import { Contract } from 'ethers';
+import ADDRESS from "@acala-network/contracts/utils/Address"
+
+import { evmChai } from '@acala-network/bodhi';
+
+const Token = require('@acala-network/contracts/build/contracts/Token.json');
```
Comment (shunjizhan) [inline precompiled-token/src/getACAinfo.ts]: should we also use `import` to be more consistent?

File: precompiled-token/src/getACAinfo.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    const balance = await instance.balanceOf(await wallet.getAddress());
+
+    console.log("Token name:", name);
+    console.log("Token symbol:", symbol);
+    console.log("Token decimal spaces:", decimals);
+    console.log("Token total supply:", totalSupply.toString());
+    console.log("Token balance of %s is: %s", walletAddress, balance.toString());
+
+    const formattedSupply = balanceFormatting(totalSupply.toString(), decimals);
```
Comment (shunjizhan) [inline precompiled-token/src/getACAinfo.ts]: `ethers` lib has a util to do this, reference: https://github.com/AcalaNetwork/bodhi.js/blob/master/eth-rpc-adapter/src/__tests__/e2e/endpoint.test.ts#L506

File: precompiled-token/src/setup.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import {
```
Comment (shunjizhan) [inline precompiled-token/src/setup.ts]: can we merge this PR and import it from util directly? https://github.com/AcalaNetwork/waffle-tutorials/pull/2/files#diff-7583809507a13391057c3aee722e422d50d961a87e2a3dbf05ea492dc6465c94R21

File: precompiled-token/test/PrecompiledToken.test.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import { expect, use } from 'chai';
+import { solidity } from 'ethereum-waffle';
+import { Contract } from 'ethers';
+import ADDRESS from "@acala-network/contracts/utils/Address"
+
+import { evmChai, Signer, TestProvider } from '@acala-network/bodhi';
+import { WsProvider } from '@polkadot/api';
+
+const Token = require('@acala-network/contracts/build/contracts/Token.json');
```
Comment (shunjizhan) [inline precompiled-token/test/PrecompiledToken.test.ts]: here also, should we also use `import`, so we won't mix cjs and esm

File: precompiled-token/test/PrecompiledToken.test.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+use(solidity);
+use(evmChai);
+
+const provider = new TestProvider({
```
Comment (shunjizhan) [inline precompiled-token/test/PrecompiledToken.test.ts]: here we can also reuse the `getTestProvider` from this PR, to be more DRY and so CI can change the URL consistently https://github.com/AcalaNetwork/waffle-tutorials/pull/2/files#diff-7583809507a13391057c3aee722e422d50d961a87e2a3dbf05ea492dc6465c94R8

File: precompiled-token/src/getACAinfo.ts (hunk 1 above)
Comment (ThunderDeliverer) [reply precompiled-token/src/getACAinfo.ts]: Sure, I don't have the preference and could find arguments for both (being concise vs. showing different options), It seems that you have a preference, so I changed this :ok_hand:

File: precompiled-token/src/getACAinfo.ts (hunk 2 above)
Comment (ThunderDeliverer) [reply precompiled-token/src/getACAinfo.ts]: It was a different util than this one, but it was a good suggestion. Thank you! :raised_hands:

File: precompiled-token/src/setup.ts (hunk 3 above)
Comment (ThunderDeliverer) [reply precompiled-token/src/setup.ts]: Done :white_check_mark:

**Comment by ThunderDeliverer** [comment]
@shunjizhan would you mind verifying that this tutorial doesn't conflict with our CI, as it has no `build` script?

**Comment by shunjizhan** [comment]
Just checked it, and it is already compatible with the CI interface (so CI should be able to run these examples), but for the tests I still got a these `Some("NoPermission")` errors when running with `mandala-node:95c483d7`. Are they already passing with new mandala versions that don't have docker version yet?

I think we can either merge them, then fix tests, or wait for the tests to be fixed and merge them together. Basically if `./run.sh test` all passed then we are all good

**Comment by ThunderDeliverer** [comment]
> Just checked it, and it is already compatible with the CI interface (so CI should be able to run these examples), but for the tests I still got a these `Some("NoPermission")` errors when running with `mandala-node:95c483d7`. Are they already passing with new mandala versions that don't have docker version yet?
> 
> I think we can either merge them, then fix tests, or wait for the tests to be fixed and merge them together. Basically if `./run.sh test` all passed then we are all good

The test pass for me with dockerized local network.. Are you using the latest version? I just pulled the repository to make sure my version of bodhi.js is up to date

**Pull request merged**